pkg/agent/plugin/svidstore/gcpsecretmanager/gcloud.go

```diff
@@ -0,0 +1,336 @@
+}
+
+type SecretManagerPlugin struct {
+	svidstorev1.UnsafeSVIDStoreServer
+	configv1.UnsafeConfigServer
+
+	log                 hclog.Logger
+	mtx                 sync.RWMutex
+	secretManagerClient secretManagerClient
+	tdHash              string
+
+	hooks struct {
+		newSecretManagerClient func(context.Context, string) (secretManagerClient, error)
+	}
+}
+
+func (p *SecretManagerPlugin) SetLogger(log hclog.Logger) {
+	p.log = log
+}
+
+// Configure configures the SecretMangerPlugin.
+func (p *SecretManagerPlugin) Configure(ctx context.Context, req *configv1.ConfigureRequest) (*configv1.ConfigureResponse, error) {
+	// Parse HCL config payload into config struct
+	config := &Configuration{}
+	if err := hcl.Decode(config, req.HclConfiguration); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "unable to decode configuration: %v", err)
+	}
+
+	if len(config.UnusedKeys) != 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "unknown configurations detected: %s", strings.Join(config.UnusedKeys, ","))
+	}
+
+	secretMangerClient, err := p.hooks.newSecretManagerClient(ctx, config.ServiceAccountFile)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to create secretmanager client: %v", err)
+	}
+
+	// gcp secret manager does not allow ".", hash td as label
+	tdHash := sha1.Sum([]byte(req.CoreConfiguration.TrustDomain)) //nolint: gosec // We use sha1 to hash trust domain names in 128 bytes to avoid secret label restrictions
+
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	p.secretManagerClient = secretMangerClient
+	p.tdHash = hex.EncodeToString(tdHash[:])
+
+	return &configv1.ConfigureResponse{}, nil
+}
+
+// PutX509SVID puts the specified X509-SVID in the configured Google Cloud Secrets Manager
+func (p *SecretManagerPlugin) PutX509SVID(ctx context.Context, req *svidstorev1.PutX509SVIDRequest) (*svidstorev1.PutX509SVIDResponse, error) {
+	opt, err := optionsFromSecretData(req.Metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get secret, if it does not exist, a secret is created
+	secret, secretFound, err := getSecret(ctx, p.secretManagerClient, opt.secretName(), p.tdHash)
+	if err != nil {
+		return nil, err
+	}
+
+	// Secret not found, create it
+	if !secretFound {
+		secret, err = p.secretManagerClient.CreateSecret(ctx, &secretmanagerpb.CreateSecretRequest{
+			Parent:   opt.parent(),
+			SecretId: opt.name,
+			Secret: &secretmanagerpb.Secret{
+				Replication: &secretmanagerpb.Replication{
+					Replication: &secretmanagerpb.Replication_Automatic_{
+						Automatic: &secretmanagerpb.Replication_Automatic{},
+					},
+				},
+				Labels: map[string]string{
+					"spire-svid": p.tdHash,
+				},
+			},
+		})
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to create secret: %v", err)
+		}
+		p.log.With("secret_name", secret.Name).Debug("Secret created")
+	}
+
+	if opt.roleName != "" && opt.serviceAccount != "" {
+		if !secretFound {
```

Comment:
I think there might be some ways to simplify this logic a bit .. maybe something like this?

```suggestion
                outdatedPolicy, err := p.shouldSetPolicy(ctx, secret, opt)
                if err != nil {
                     return nil, err
                }
                
                if !secretFound || outdatedPolicy {
                     p.setIamPolicy ...
```